openstack/standalone-cinder/pkg/provisioner/provisioner.go

```diff
@@ -0,0 +1,201 @@
+
+	// CinderVolumeID is an annotation to store the ID of the associated cinder volume
+	CinderVolumeID = "cinderVolumeId"
+)
+
+type cinderProvisioner struct {
+	// Openstack cinder client
+	VolumeService *gophercloud.ServiceClient
+
+	// Kubernetes Client. Use to create secret
+	Client kubernetes.Interface
+	// Identity of this cinderProvisioner, generated. Used to identify "this"
+	// provisioner's PVs.
+	Identity string
+
+	vsb volumeServiceBroker
+	mb  mapperBroker
+}
+
+// NewCinderProvisioner returns a Provisioner that creates volumes using a
+// standalone cinder instance and produces PersistentVolumes that use native
+// kubernetes PersistentVolumeSources.
+func NewCinderProvisioner(client kubernetes.Interface, id, configFilePath string) (controller.Provisioner, error) {
+	volumeService, err := volumeservice.GetVolumeService(configFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get volume service: %v", err)
+	}
+
+	return &cinderProvisioner{
+		VolumeService: volumeService,
+		Client:        client,
+		Identity:      id,
+		vsb:           &gophercloudBroker{},
+		mb:            &volumeMapperBroker{},
+	}, nil
+}
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+	var (
+		connection volumeservice.VolumeConnection
+		mapper     volumeMapper
+		pv         *v1.PersistentVolume
+		cleanupErr error
+	)
+
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	volumeID, err := p.vsb.CreateCinderVolume(p.VolumeService, options)
+	if err != nil {
+		glog.Errorf("Failed to create volume")
+		goto ERROR
+	}
+
+	err = p.vsb.WaitForAvailableCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Volume %s did not become available", volumeID)
+		goto ERROR_DELETE
+	}
+
+	err = p.vsb.ReserveCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Failed to reserve volume %s: %v", volumeID, err)
+		goto ERROR_DELETE
+	}
+
+	connection, err = p.vsb.ConnectCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Failed to connect volume %s: %v", volumeID, err)
+		goto ERROR_UNRESERVE
+	}
+
+	mapper, err = p.mb.newVolumeMapperFromConnection(connection)
+	if err != nil {
+		glog.Errorf("Unable to create volume mapper: %f", err)
+		goto ERROR_DISCONNECT
+	}
+
+	err = mapper.AuthSetup(p, options, connection)
+	if err != nil {
+		glog.Errorf("Failed to prepare volume auth: %v", err)
+		goto ERROR_DISCONNECT
+	}
+
+	pv, err = p.mb.buildPV(mapper, p, options, connection, volumeID)
+	if err != nil {
+		glog.Errorf("Failed to build PV: %v", err)
+		goto ERROR_DISCONNECT
+	}
+	return pv, nil
+
+ERROR_DISCONNECT:
+	cleanupErr = p.vsb.DisconnectCinderVolume(p.VolumeService, volumeID)
+	if cleanupErr != nil {
+		glog.Errorf("Failed to disconnect volume %s: %v", volumeID, cleanupErr)
```

Comment:
I don't know the best answer but later in `Delete()`, similar error leaves rest of the code unexecuted. Some consistency is preferred

Reply:
I am not sure the best answer here either.  If something is wrong with cinder or the storage backend the cleanup actions aren't too likely to succeed either.  I've been deferring a decision on how to handle this until I see more actual scenarios in the wild.  I can change the Delete flow to work the same as the Provision cleanup path if that makes more sense.